OpenGL improvements

binary1248 opened the pull request:
It's done :grin:.

I was too lazy to copy all the commit messages here, so if you have any questions feel free to ask.

Needs testing on OS X.

LaurentGomila commented:
Yeah :+1:

Wouldn't it be clearer to add an enum for the context flags, rather than separate booleans? It would take less arguments, and be clearer if set with the constructor.

Isn't there a license associated to the code generated by GLLoader?

Was it really necessary to request a 24-bit depth buffer in the OpenGL and Window examples? The code is supposed to pick the closest valid configuration if some attributes are not available.

eXpl0it3r commented on src/SFML/Graphics/Shader.cpp:
```diff
@@ -40,6 +40,18 @@
+    #define castToGlHandle(x) (x)
+    #define castFromGlHandle(x) (x)
+
+#endif
```
Why do you use macros here?

binary1248 replied:
Because those lines are literally no-ops. I've never seen any cases where one declares functions just to conditionally cast stuff. Then again not many people run into this problem that often either. :stuck_out_tongue:

What do you suggest?

eXpl0it3r replied:
Ah okay, yeah I guess it makes sense. :wink:

TankOs replied:
I suggest a function. Is there any issue with a function?

binary1248 replied:
What would the function look like?

TankOs replied:
```cpp
template <class T>
GLEXT_GLhandle castToGlHandle(T handle) {
  return reinterpret_cast<GLEXT_GLhandle>(static_cast<ptrdiff_t>(handle));
}

template <>
GLEXT_GLhandle castToGlHandle(GLEXT_GLhandle handle) {
  return handle;
}
```

Might need type adjustments, haven't checked the original source.

binary1248 replied:
But in this case, template argument deduction won't work.

We need something that takes an `unsigned int` and returns a `GLEXT_GLhandle` in both cases. The difference is not the parameter type (which is always `unsigned int`), but the return type (which can be `unsigned int` or `void*`). I don't see any way around the preprocessor for this...

TankOs replied:
> We need something that takes an unsigned int and returns a GLEXT_GLhandle in both cases. 
>[...]
> The difference is[...]the return type 

Return GLEXT_GLhandle in both cases or not? ;)

MarioLiebisch replied:
Actually, do you even need the case differentiation at all? Can't you just use `reinterpret_cast()` in all cases, even if input and output are identical? Never tried that to be honest, but don't see why it shouldn't work.

LaurentGomila replied:
> But in this case, template argument deduction won't work.

It does. Since T is unsigned int, the following happens:
* If GLEXT_GLhandle is unsigned int, then the specialization is used and no conversion happens
* if GLEXT_GLhandle is void* then the generic implementation is used and the conversion is done

binary1248 replied:
> It does. Since T is unsigned int, the following happens:

Tank provided an example for `castFromGlHandle()` but named it wrong and had the wrong return type (should be `unsigned int`).

You guys forget that this is the `castToGlHandle()` function. :wink: It has to cast from a fixed type (`unsigned int`) in both cases to whatever `GLEXT_GLhandle` is. `castFromGlHandle()` casts from a `GLEXT_GLhandle` to `unsigned int` in both cases. So yeah... your template solution only works for one direction but not the other.

> Actually, do you even need the case differentiation at all? Can't you just use reinterpret_cast() in all cases, even if input and output are identical? Never tried that to be honest, but don't see why it shouldn't work.

reinterpret_cast() doesn't allow casting from a pointer to an integral value when the conversion would truncate the data (i.e. 64-bit pointer -> 32-bit `unsigned int`). I tried this already. Apple really tried their hardest to be as annoying as possible. :wink: But yeah... I guess the inner static_cast in the "To" case can be removed. But moving the double cast into every line where it is needed would make the code really horrible to read IMHO.

I'm still interested to see suggestions for how to cast from an `unsigned int` to `GLEXT_GLhandle` that works in both Apple and non-Apple cases :grin:.

TankOs replied:
https://gist.github.com/TankOs/d89f0d58fe48367bb5bd

Maybe you can even get rid of the still remaining PP directive, haven't checked the original code.

binary1248 replied:
:grin:

Like I said, `castToGlHandle` has to have the signature
```cpp
GLEXT_GLhandle castToGlHandle(unsigned int)
```
in both cases. Your gist still doesn't do this. Those macros aren't for casting from `void*` to anything else or back. They are only there to cast from `unsigned int` to `GLEXT_GLhandle` and vice versa. It's the `GLEXT_GLhandle` type that changes and not the function signature.

In essence, you would have to have 2 functions both with the same name and taking the same parameter type but returning a different type depending on system. Like I said, I don't think this is doable with templates alone.

LaurentGomila replied:
I vote for the macros. We're wasting so much time overengineering this, just for the beauty of C++... Let's be pragmatic and try to make more useful comments on this huge PR :stuck_out_tongue:

TankOs replied:
Ahhhh, got it now. Macro seems legit then. :smirk_cat:

binary1248 commented:
> Wouldn't it be clearer to add an enum for the context flags, rather than separate booleans? It would take less arguments, and be clearer if set with the constructor.

Perhaps... didn't think of that all this time somehow :stuck_out_tongue:.

> Isn't there a license associated to the code generated by GLLoader?

No... we use a _loader generator_ to generate the code that you see in GLLoader.cpp and GLLoader.hpp, and even that (initial) code is heavily modified. It's like using MS Paint to draw a picture and asking if that picture has to have a specific license :stuck_out_tongue:. I don't see any problems here.

> Was it really necessary to request a 24-bit depth buffer in the OpenGL and Window examples? The code is supposed to pick the closest valid configuration if some attributes are not available.

If you try to request settings that don't perfectly match what the system can provide, you will stumble into the fallback path. On some systems, that might lead to other settings not being met as well, going as far as decreasing the OpenGL version number to < 3.0 because the initial attempt at getting a 3.0+ didn't yield any matching configurations. Sure, it isn't _that_ important in the examples, but I figured since they are examples, why not make them able to run on more systems without having to go into the fall back path :stuck_out_tongue:.

There currently isn't a way for the user to enumerate all the valid configurations that the system can provide, so if a certain setting is important to them, they just have to keep trying all possible combinations until nothing gets "downgraded". This happens even when settings are completely unrelated. This might be something we would want to add in the future in addition to the video mode enumeration.

mantognini commented:
> Needs testing on OS X.

Alright. I'll do it probably in 10 days (after the exams). Beside code snippet from #778 and the examples, is there anything specific I should test? Any corner cases?

binary1248 commented:
#778 occurred on Windows, or at least I encountered it only on Windows. It might not be an issue on OS X even with the current master.

As for testing the rest, no there isn't anything specific to test. If the pong, shader and OpenGL examples build and run fine (without any warnings as well) then it is more or less safe to assume that it "just works" on OS X.

mantognini commented:
:ok_hand: I'll test that.

binary1248 commented:
The new version provides the context attribute flags as enums.

Anything else? :grin:

LaurentGomila commented:
Flags that can be combined use the ```Uint32``` type in other classes (```sf::Window```, ```sf::Text```), so... :grimacing:

binary1248 commented:
Anything else? :grin:

mantognini commented:
> As for testing the rest, no there isn't anything specific to test. If the pong, shader and OpenGL examples build and run fine (without any warnings as well) then it is more or less safe to assume that it "just works" on OS X.

Pong and shader run fine. However, the OpenGL sample is broken: there is no longer the cube in place of the cursor. Is it the same on Linux/Windows or yet another OS X speciality?

binary1248 commented:
I always test all examples on all the systems I have (Windows and Linux). The OpenGL example works on both for me. If you can, try to get a console up and look for errors. It might have something to do with the context settings/versions.

mantognini commented:
Nothing in the terminal nor the logger. Using git bitsect I can incriminate 1dae59c1143e6cece00895a9e9b5303081a018ee, if that helps you...

binary1248 commented:
The "strange" thing is that according to your description, only the rotating cube is missing. Since the background and text are still rendered properly, sfml-graphics and sfml-window have to work as intended in regards to how they take care of context management. The problem has to be specific to the OpenGL code (or build configuration settings?) in the example since it doesn't occur with the the pong or shader examples.

mantognini commented:
I've to admit I don't understand everything that's going on. For example, I see `Unix/GlxExtensions.*` and `Win32/WglExtensions.*` but nothing similar for the other platforms, is it intentional?

I've enable more warnings at compilation, I get a lot of this:

> ```
SFML/src/SFML/Graphics/GLLoader.cpp:56:34: warning: cast between
      pointer-to-function and pointer-to-object is an extension [-Wpedantic]
    sf_ptrc_glBlendEquationEXT = (void (CODEGEN_FUNCPTR *)(GLenum))IntGetProcAddress("glBlendEquationEXT");
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

No idea if it matters, though.

BTW, I also get this one, not that it really matters.
> ```
SFML/include/SFML/Window/ContextSettings.hpp:46:25: warning: commas at the end of
      enumerator lists are a C++11 extension [-Wc++11-extensions]
        Debug   = 1 << 2, ///< Debug attribute
```

Regarding those two flags (Debug and Core), the doc should probably state that they do nothing on OS X.

binary1248 commented:
> For example, I see Unix/GlxExtensions.* and Win32/WglExtensions.* but nothing similar for the other platforms, is it intentional?

Yes... OS X doesn't rely on an extension to get access to the newer context management functions.

I'll look into those warnings. They shouldn't really change anything though, since pong and shader run fine. You do see the background and text in the OpenGL example right? Just not the spinning cube? Also, does it stop working right after the commit that introduces GLLoader, or only after a later commit?

mantognini commented:
> You do see the background and text in the OpenGL example right? Just not the spinning cube?

Exactly, I see only background and text.

> Also, does it stop working right after the commit that introduces GLLoader, or only after a later commit?

Commit on `bugfix/glx_context` are fine, the first with this issue is the one introducing the loader, that is, after your rebase, this one: https://github.com/LaurentGomila/SFML/commit/dab8ecc3b0bf30d92875fc5e2f7788dea0f641b5

binary1248 commented:
Hmm... I'm running out of ideas :confused:.

Can you try and replace the rotating cube by something "simpler", like a simple untextured triangle using glBegin() and glEnd()? It might be one of those cases where the code relies on... "not so well specified behaviour" such as the starting OpenGL state which might be different on different implementations.

Also... it would be nice if others with OS X can test as well to see if this is "reproducible" on their systems as well.

mantognini commented:
I tried a few things, such as

```cpp
        glColor3f(1.0, 1.0, 1.0);
        glMatrixMode(GL_PROJECTION);
        glLoadIdentity();
        glOrtho(0.0, 1.0, 0.0, 1.0, -1.0, 1.0);
        glBegin(GL_POLYGON);
        glVertex3f(0.25, 0.25, 1.0);
        glVertex3f(0.75, 0.25, 1.0);
        glVertex3f(0.75, 0.75, 0.0);
        glVertex3f(0.25, 0.75, 0.0);
        glEnd();
```

which works fine.

And by luck, I ended up having something close to working. [In this gist](https://gist.github.com/mantognini/54e4527d92e5bed76486) you can find a edited version of the OpenGL example where basically I moved all ogl code inside the main loop (except for the texture loading). Here is the result:

![ugly](https://cloud.githubusercontent.com/assets/693483/5940739/1a8c44ca-a70d-11e4-89f0-64c0c31afe3e.png)

Instead of no cube I got this brownish thing...

Does it give you any idea of where I should look next?

binary1248 commented:
Hmm... somehow the texture isn't being applied. What happens if you move the texture creation block inside the loop (preferably after enabling `GL_TEXTURE_2D`) as well? You will want to delete the texture object at the end of the loop body if you do this...

binary1248 commented:
:point_up: :grin:

mantognini commented:
Sorry, I didn't see your previous message...

So, just moving [L50-L60](https://gist.github.com/mantognini/54e4527d92e5bed76486#file-opengl_edited-cpp-L50-L60) inside the main loop doesn't improve things. Now I remembered you mentioning `resetGLStates()` on the forum recently. So I tried adding `window.resetGLStates();` right before loading ([see this other gist](https://gist.github.com/mantognini/bbedd87bf8cff4f05cf5#file-opengl_edited2-cpp-L117)) and it worked. :-)

I tried to do the same and add `window.resetGLStates();` before [line 50 in the first gist](https://gist.github.com/mantognini/54e4527d92e5bed76486#file-opengl_edited-cpp-L50) and it worked too.

If I do it on [line 49 of the opengl example](https://github.com/LaurentGomila/SFML/blob/feature/gl_dev_new/examples/opengl/OpenGL.cpp#L49), it works as well!

Do you have any clue why it behaves like that?

binary1248 commented:
Well... the next step would be to replace `window.resetGLStates();` with the code that it runs :stuck_out_tongue:.

Place the following block where `window.resetGLStates();` is ([line 49 of the opengl example](https://github.com/LaurentGomila/SFML/blob/feature/gl_dev_new/examples/opengl/OpenGL.cpp#L49)) and see if it works.
```cpp
sf::Shader::isAvailable();
window.setActive();

glClientActiveTexture(GL_TEXTURE0);
glActiveTexture(GL_TEXTURE0);

glDisable(GL_CULL_FACE);
glDisable(GL_LIGHTING);
glDisable(GL_DEPTH_TEST);
glDisable(GL_ALPHA_TEST);
glEnable(GL_TEXTURE_2D);
glEnable(GL_BLEND);
glMatrixMode(GL_MODELVIEW);
glEnableClientState(GL_VERTEX_ARRAY);
glEnableClientState(GL_COLOR_ARRAY);
glEnableClientState(GL_TEXTURE_COORD_ARRAY);

glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA, GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
glBlendEquationSeparate(GL_FUNC_ADD, GL_FUNC_ADD);
glLoadIdentity();
sf::Texture::bind(NULL, sf::Texture::Pixels);
sf::Shader::bind(NULL);

window.setView(window.getView());
```
If it does work, try to remove 1 line at a time until it doesn't work any longer. Hopefully we will end up with 1 or 2 lines that are responsible for "fixing" the problem.

eXpl0it3r commented:
I'll just jump in to say, that the OpenGL example works fine for me on Windows 8.1 with my AMD Radeon HD 7500M/7600M.

mantognini commented:
So, apparently, the `sf::Shader::isAvailable` + `window.setActive` duo workarounds the issue. But when I tried to replace isAvailable by its code that has side effects, I couldn't make the texture work.

Do you see any reason why, in this specific piece of code, `sf::Shader::isAvailable` is not equal to the next block?
```cpp
{
    sf::Context ctx;
    sf::priv::ensureExtensionsInit();
}
```

([full code](https://gist.github.com/mantognini/ed76e60a37622fbb0a5e))

binary1248 commented:
Hmm... that does seem strange.

I guess the only way to tell if _exactly_ the same "stuff" happens is to step in each variant with a debugger and compare what is done at an even lower level. A lot of the stuff underneath relies heavily on internal state, and they might not be the same even though the code has one think that.

mantognini commented:
I'll try to dig into it next week -- can't do it sooner. :-/

eXpl0it3r commented:
Since I don't get it on my AMD card on Windows I wonder, is this issue specific to OS X or to a certain GPU vendor?

mantognini commented:
If someone with a NVIDIA wants to test it on Linux/Window that could rule out, or not, an OS X specific issue. (I've got a NVIDIA GeForce GT 650M btw.)

binary1248 commented:
Might be a long shot, but the way OS X does things could be a bit... different from other operating systems.

One of the main differences between the new implementation and the old one is that the new one relies on manually loading _all_ OpenGL entry points, even the ones that the standard OpenGL library (e.g. the 1.1 one on Windows) would have provided anyway.

Since the OpenGL example is using the system's OpenGL header (and thus uses the entry points from the system's default OpenGL library) and [SFML internally manually loads the entry points from _some_ OpenGL dynamic library](https://github.com/LaurentGomila/SFML/blob/feature/gl_dev_new/src/SFML/Window/OSX/SFContext.mm#L116), it could very well be that they end up calling functions from different libraries. This was never a problem on Windows or Linux because they tend to have just a single OpenGL library that you can use, but OS X might have multiple (I'm just guessing here). Since context state is probably local to the library whose functions you are calling, the assumption that state changes that SFML makes apply to the example's GL environment doesn't hold any more.

I don't know how you would go about doing this, but maybe you can figure out a way to check what the destination of the GL calls inside SFML and in the example are. If they are indeed different, I'll have to come up with something special for our friend OS X :wink:.

mantognini commented:
Ho God! :weary: 

I'll see what I can do but it might be a while until I get something useful...

AlexAUT commented:
I tried the opengl example with the gl_dev_new branch with an NVIDIA gt740m under windows and it renders fine, but NVIDIA NSIGHT is telling my about an invalid operation. I guess it's because of legacy opengl, anyways here is the log from NSIGHT: http://pastebin.com/YDq5Gm8S

binary1248 commented:
Yeah those are all legacy warnings. CodeXL spits them out too. They can be ignored.

binary1248 commented:
@jcowgill: Thanks for the error report. It seems that the Intel part of your optimus setup doesn't like XSetWindowColormap for some reason. I'm not sure if it is even required since we are rendering directly through OpenGL.

Does it work if you comment out lines [355](https://github.com/LaurentGomila/SFML/blob/feature/gl_dev_new/src/SFML/Window/Unix/GlxContext.cpp#L355) to [363](https://github.com/LaurentGomila/SFML/blob/feature/gl_dev_new/src/SFML/Window/Unix/GlxContext.cpp#L363) in GlxContext.cpp of the latest commit?

jcowgill commented:
Yes it seem to work fine if I comment that out.

binary1248 commented:
Hmm... yeah I just did a bit of reading and it seems that the X colormap is used by OpenGL in color index mode rendering. Since nobody really used it, it was deprecated and eventually removed from the core OpenGL profile.

So... I guess those lines can go :stuck_out_tongue:.

binary1248 commented:
@jcowgill: The change should be pushed now. It should work on both adapters in your Optimus set up.

binary1248 commented:
@mantognini: Can you try out the [branch without dynamic 1.1 loading](https://github.com/LaurentGomila/SFML/tree/feature/gl_dev_no_1_1)? It is almost identical to how GLEW does it, which obviously works on OS X as well.

mantognini commented:
I've got the same results with both branches. :-/

binary1248 commented:
@mantognini: And [this branch](https://github.com/LaurentGomila/SFML/tree/feature/gl_dev_no_recreate)?

mantognini commented:
It seems to work fine. :smiley:

binary1248 commented:
Rebased onto master.

So... I guess the final testing and code review phases can start :stuck_out_tongue:.

LaurentGomila commented:
So what was wrong?

mantognini commented:
Tested again for safety on Mac, this branch works fine! Good work there @binary1248

binary1248 commented:
The previous implementation tried to recreate all existing contexts with the settings of the context that is currently created (shared context = legacy, new context = core => shared context core as well). This didn't work out with OS X because of a few differences in how contexts are created there as opposed to in GLX and Wgl, e.g. it doesn't create a new window/view for every single context.

Even though in theory it should have worked, _something_ went wrong and I just resorted to the previous behaviour of letting contexts of different settings be created and shared with each other. This might not always work (the spec makes no guarantees), but testing so far hasn't shown any problems sharing core and compatibility contexts with/without debug bit set.

It is also probably vendor dependent how debug contexts work. If functions are actually executed elsewhere internally (in a non-debug context) even through we issued it in a debug context it might not trigger any debugging related stuff. Since this is not standardized, it doesn't make sense to enforce anything ourselves, if it won't work on a certain implementation it won't work.

mantognini commented on include/SFML/OpenGL.hpp:
```diff
@@ -33,7 +33,7 @@
 
 ////////////////////////////////////////////////////////////
-/// This file just includes the OpenGL (GL and GLU) headers,
+/// This file just includes the OpenGL headers,
```
Is there an issue to use glu in the user code? If not, I'd still include them for convenience.

binary1248 replied:
All GLU functionality can be reproduced with pure GL and user code. Think of them like... "macros" that call some very very ancient GL code. It is widely established (even since before modern GL became a thing) that GLU should be avoided whenever possible, which is almost always.

It is not guaranteed that GLU will even be available on all platforms. Some (look at the mobile platforms) might have even thrown it out already.

mantognini replied:
Then I'm happy with the changes. :-)

mantognini commented on include/SFML/Window/ContextSettings.hpp:
```diff
@@ -94,6 +109,23 @@ struct ContextSettings
+/// additional debugging features enabled. Depending on the
+/// system, this might be required for advanced OpenGL debugging.
+/// OpenGL debugging is disabled by default.
+///
```
Regarding core/compatiliby flag with version >= 3.2 and debug flag, we should probably add a note about OS X. Something along those lines:

<b>Special Note for OS X:</b> Apple only provide a legacy profile (OpenGL 2.1) and a core profile for some versions of the operating system (further information are available on the <a href="https://developer.apple.com/opengl/capabilities/index.html">OpenGL Capabilities Tables</a> page). Moreover, no debug context are available on that system.

```
<b>Special Note for OS X:</b> Apple only provide a legacy profile (OpenGL 2.1) and a core profile for some versions of the operating system (further information are available on the <a href="https://developer.apple.com/opengl/capabilities/index.html">OpenGL Capabilities Tables</a> page). Moreover, no debug context are available on that system.
```

mantognini commented on src/SFML/Graphics/GLExtensions.hpp:
```diff
@@ -34,80 +34,175 @@
+    #define GLEXT_GL_FRAMEBUFFER_BINDING              GL_FRAMEBUFFER_BINDING_OES
+    #define GLEXT_GL_INVALID_FRAMEBUFFER_OPERATION    GL_INVALID_FRAMEBUFFER_OPERATION_OES
+
+#else
```
With nested `#if/else` I'd recommend adding a comment mentioning line 33 or SFML_OPENGL_ES being not set.

mantognini commented on include/SFML/Graphics/Shader.hpp:
```diff
@@ -457,6 +457,18 @@ class SFML_GRAPHICS_API Shader : GlResource, NonCopyable
+    /// \return OpenGL handle of the shader or 0 if not yet loaded
+    ///
+    ////////////////////////////////////////////////////////////
+    unsigned int getNativeHandle() const;
```
I'm wondering if the cast macros from `src/SFML/Graphics/Shader.cpp` should be documented here. Having never use those handle myself I don't know whether or not the user will have some trouble with Apple's special types...

binary1248 replied:
We can only return a single type in the interface, unless we expose the GLLoader header, which we don't want.

Theoretically the user should be able to cast from `unsigned int` to whatever handle type they are using. The compiler will complain a lot if you aren't careful, but other than that it shouldn't really be a problem. The people who know how to use the handle will be able to take care of this themselves, after all the documentation already discourages people who don't know what they are doing from using this anyway :wink:.

Side note: The casting is really only a problem in legacy contexts on OS X because shaders are an extension. In core contexts (with shaders as part of the core specification), the OpenGL specification specifies the type of the handle for us (which is the same as all the other handles), so Apple isn't able to be creative any more :stuck_out_tongue:.

mantognini replied:
>  after all the documentation already discourages people who don't know what they are doing from using this anyway

Indeed.

mantognini commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -421,6 +421,18 @@ class SFML_GRAPHICS_API Texture : GlResource
+    /// \return OpenGL handle of the texture or 0 if not yet created
+    ///
+    ////////////////////////////////////////////////////////////
+    unsigned int getNativeHandle() const;
```
Same remark as for `Shader::getNativeHandle()`.

binary1248 replied:
Unlike with the shaders, on all systems (including OS X), the handle type for textures is an integer value (specified in OpenGL specification), so no casting needed here.

mantognini commented on src/SFML/Window/GlContext.cpp:
```diff
@@ -73,9 +68,38 @@
 
 #endif
 
+#if !defined(GL_MULTISAMPLE)
```
Is there a specific reason why those macros are defined here and not in `GLLoader.hpp`?

binary1248 replied:
GLLoader.hpp takes care of the OpenGL functions and definitions (the "rendering stuff"). GL_MULTISAMPLE and all the stuff underneath are related to context management and/or window/pixel format specifications. They aren't supposed to be defined together. This is the way it is in the official headers from Khronos as well. Previously they were in glext.h (not GL.h), but I got rid of that because we don't need a 10000 line file just for a few definitions :wink:.

mantognini replied:
Alright then. :-)

mantognini commented on src/SFML/Window/OSX/SFContext.mm:
```diff
@@ -183,32 +199,48 @@
+        m_settings.majorVersion = 3;
+        m_settings.minorVersion = 2;
         attrs.push_back(NSOpenGLPFAOpenGLProfile);
         attrs.push_back(NSOpenGLProfileVersion3_2Core);
```
I've just read something saying that this flag might create a 4.x context on 10.9+ on some hardware. This implies that L291 might report slightly wrong information. But we could just close our eyes and pretend Apple never used a ***3_2*** flag for 4.x context...

binary1248 replied:
I guess they were too lazy to think of this... :wink:

Regardless, the OpenGL context version is updated in GLContext.cpp anyway, so it should always contain the true context version if OpenGL reports it properly.

The check at L291 also only triggers if the actual version is less than the requested version, which in this case is not true (4.x > 3.2).

mantognini replied:
> The check at L291 also only triggers if the actual version is less than the requested version

Or if he asks for a debug context. ;-)

binary1248 replied:
But that part is working as intended :wink:.

binary1248 commented:
I moved the checks into GlContext, indented the definitions in GlExtensions.hpp and added the OS X note to ContextSettings.hpp.

Tell me if there are more issues.

mantognini commented:
No more question/comment for me. NB: I haven't looked into Unix/Windows part.

Enabling compiler warnings gives me the same as [before](https://github.com/LaurentGomila/SFML/pull/779#issuecomment-71051411) though. The second one is easily fixed (just remove the comma). Regarding the casting warning, it's a bit more confusing. This [SO question](http://stackoverflow.com/questions/14125474/casts-between-pointer-to-function-and-pointer-to-object-in-c-and-c) could be a beginning of an answer.

binary1248 commented:
The latest version should fix the warnings.

The cast warning was coming from the fact that the `getFunction()` methods were returning a function pointer as a simple `void*` (which is a pointer-to-object). They now return `sf::GlFunctionPointer` instead which is a typedef to `void (*GlFunctionPointer)()`.

mantognini commented:
Perfect for me! :-)

binary1248 commented:
I merged the glx_swap_interval branch into this one and repositioned its commit since half of the fix for #727 was already part of gl_dev. This PR should be complete now, i.e. I have nothing more to add :wink:.

mantognini commented:
Still working fine for me.

NB: if this completely supersedes #727 then don't forget to add the proper tags and close the issue. :wink:

binary1248 commented:
I like keeping the original issues open until they are truly fixed in master. :wink:

eXpl0it3r commented:
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

mantognini commented:
I'm wondering if it was tested on Android/iOS? (I guess it will work but we never know...)

Also, it would be a good idea if a third pair of eyes could have a look at the commits too. :-)

binary1248 commented:
If you look at the changes, they don't touch anything Android/iOS related. GLEW was never used there... Context management was left untouched... Shaders haven't been implemented... The examples were refactored, nothing more.

mantognini commented:
Nevertheless, at least checking if it compiles shouldn't take too long and can only be beneficial. :wink:

eXpl0it3r commented:
You have a Mac, so feel free to test iOS. :stuck_out_tongue_winking_eye:

binary1248 commented:
Actually... I think I might have forgotten to implement sf::Shader::getNativeHandle() for OpenGL ES :fearful:.

Going to fix now...

binary1248 commented:
There.. should be fixed now.

Any volunteers to test on Android/iOS? :grin:

BlueCobold commented:
Sure thing, I can test for OSX and iOS. eXpl0it3r just informed me about this PR. Do I understand correctly that Shaders are now supported on mobile platforms?

LaurentGomila commented:
> Do I understand correctly that Shaders are now supported on mobile platforms?

Nop. And that's not going to happen soon, because it requires OpenGL ES 2, which would involve to rewrite the whole graphics module.

binary1248 commented:
> which would involve to rewrite the whole graphics module.

Don't exaggerate :wink:. Maybe like... 20% of the graphics module would have to be "adapted" to not use legacy OpenGL and ES 2 support would come for free. The first step on the way there is this PR obviously :stuck_out_tongue:.

binary1248 commented:
Yeah... so I just found out that XCB has it's [own version of GLX](http://xcb.freedesktop.org/manual/group__XCB__Glx__API.html) that can be used instead of the "classic" GLX functions. The main advantage is that we wouldn't have to set up some elaborate error handling system (as would be necessary using Xlib, but SFML never bothered to do previously either). Further reading [here](http://www.remlab.net/op/xlib.shtml).

The reason why many people suddenly crash out of their application is because X might generate an error in some GLX calls that don't get handled anywhere since we are using XCB now. All other libraries that I have looked at still use Xlib and handle GLX errors through XSetErrorHandler. This catches errors like those that are produced if we feed invalid context versions to glxCreateContextAttribsARB (yes... the GLX specification says it has to raise an X error, which when unhandled just closes the application with some debug information). We obviously can't afford to do this any more, so there will be no other choice than to use XCB GLX.

This will need a bit of rewriting :stuck_out_tongue:.

binary1248 commented:
Actually, I just found out that if one were to go through XCB GLX, it would end up using indirect rendering, which we absolutely don't want. I'm probably going to fix up the error handling using Xlib/XCB interoperation for now (this is really the only way), and in the future we can think about Wayland support along with OpenGL ES on Unix systems as well. It seems that X is slowly dying out and XCB is just a last ditch effort to prolong its lifetime. Many former X developers have already moved on to work on Wayland by now.

binary1248 commented:
The fatal error causing X to close the application has been fixed (handled/ignored), along with a few other minor bugs. Thanks to @AlexAUT for the bug report regarding the X error.

I also added optional debugging to GlxContext.cpp that is disabled by default. It might help in really mysterious cases of OpenGL context related issues.

BlueCobold commented:
Doesn't compile for iOS. GL_OES_texture_npot is unknown.

binary1248 commented:
> Also, there is some GLLoader.hpp/cpp in the graphics folder, but they aren't part of CMake/the project. What's going on with these?

They are only used to load extensions on desktop systems, where it is necessary unlike on mobile devices. And yeah... they are part of the CMake configuration, just not part of the main file listing block and excluded from projects generated for mobile systems.

> Doesn't compile for iOS. `GL_OES_texture_npot` is undefined.

Can you try replacing `GL_OES_texture_npot` with `GL_APPLE_texture_2D_limited_npot` and see if it works? Simply setting it to false would disable support on Android devices that would have supported npot textures as well. Apple decides to roll their own extension for this I guess...

BlueCobold commented:
> Can you try replacing GL_OES_texture_npot with GL_APPLE_texture_2D_limited_npot and see if it works?

GL_APPLE_texture_2D_limited_npot is an extension which must be checked for at runtime. It should never be used as a static define as for example an iPad Air supports it, but an iPhone 3G doesn't.

binary1248 commented:
I must admit, I don't have much experience developing OpenGL ES for mobile devices, so I simply assumed that the system would "make it work" if the headers define a specific interface. That was what I discerned from [what is currently in master](https://github.com/LaurentGomila/SFML/blob/master/src/SFML/Graphics/GLExtensions.hpp).

If we have to check for extensions in the same way as on the desktop, then that can be arranged as well, but your worry that something might be used on a platform which doesn't support it holds for master already, so maybe this might be taken care of in a separate issue. From [this](http://stackoverflow.com/questions/6970725/android-opengl-es-all-extensions-supported), it seems like eglGetProcAddress is only supported from android-9 onwards, and I have no idea what API SFML is currently targeting.

And about npot: [The specification](https://www.khronos.org/registry/gles/extensions/OES/OES_texture_npot.txt) states that it can be present on systems that can provide an OpenGL ES 1.1 or 2.0 context as well, so ES 2 is not the minimum requirement for npot support. The opposite is true however, ES 2.0 support implies _limited_ (and for our purposes, adequate) npot support.

BlueCobold commented:
What I meant is, that just because the constant is defined, it doesn't automatically mean that the devices supports it. In ES 2 this is a must-have-feature afaik, but in ES 1.X it isn't. Since SFML is using 1.X, generally assuming that the support exists, is a bug. You might be right that this is already from the previous SFML incarnation and thus not your mistake. It does explain though why in that header there are so many "some Android devices don't support" lines exist. Because of course they don't automatically support GL ES 2 features in an ES 1 context (or at all).
But yea, if this is a problem of a previous SFML code base, it should be addressed elsewhere. 

However, the current npot-definition doesn't compile for iOS and setting it to false or checking for the extension at runtime are the only valid options here.

binary1248 commented:
I guess we can be conservative for now and simply set npot support to `false` then.

BlueCobold commented:
I think so, yes. It would be king's league to check it at runtime, but that's a new feature imo.

binary1248 commented:
The updated branch should resolve the X visual mismatch that came around after the XCB rewrite.

eXpl0it3r commented:
Needs rebasing.

@MarioLiebisch Could you give this a try on Android?

MarioLiebisch commented:
Compiles, links and runs fine on my Nexus 5 as far as I can tell.

zsbzsb commented:
Just one minor note: there is still old copyright years and Laurent's email address in the added header files.

binary1248 commented:
Rebased onto master and updated copyright notices. I think with the exception of #803 (which - pending further information - I have to assume is caused by the Optimus setup involved), there should be no more issues with this branch.

Feel free to give it a final test run and code review. It's time to put this branch behind us. :wink:

eXpl0it3r commented:
This PR has been added to my merge list, meaning it will be merged after a few positive feedbacks, unless someone raises any concerns.

mantognini commented:
> Feel free to give it a final test run and code review.

Still working fine here. :-)

eXpl0it3r commented:
:tada: :tada:

mantognini commented:
:clap: :clap: :smiley:
